### PR TITLE
dnstable-convert 0.14.0 - Block size and thread count options

### PR DESCRIPTION
**Pull request by kadealicious**
- Add -b <size> option to specify uncompressed data block size for output files.
 - Add -t <count> option to enable multithreading (see mtbl PR #84).

### CODE REVIEW

File: dnstable_convert.c [hunk 1]
```diff
@@ -984,18 +988,24 @@ init_mtbl(mtbl_compression_type compression, int level, size_t mem_mb)
 		mtbl_sorter_options_set_temp_dir(sopt, getenv("VARTMPDIR"));
 
 	wopt = mtbl_writer_options_init();
+
+	mtbl_writer_options_set_threadpool(wopt, pool);
+	mtbl_writer_options_set_block_size(wopt, block_size);
```
Comment (djw1149) [inline dnstable_convert.c]: move line 993 to an else block after line 1000

File: dnstable_convert.c [hunk 2]
```diff
@@ -1060,6 +1075,22 @@ main(int argc, char **argv)
 				return (EXIT_FAILURE);
 			}
 			break;
+		case 'b':
+			block_size = atoi(optarg);
```
Comment (djw1149) [inline dnstable_convert.c]: please switch lines 1079 and 1087 from using atoi to use strtol code from lines 1071...1072

File: dnstable_convert.c [hunk 3]
```diff
@@ -100,7 +101,8 @@ static struct timespec		start_time;
 static uint64_t			count_messages;
 static uint64_t			count_entries_dns;
 static uint64_t			count_entries_dnssec;
-static uint64_t			count_entries_merged;
+/* Count merged entries correctly even when MTBL multithreading is enabled. */
+static atomic_uint_fast64_t	count_entries_merged;
```
Comment (djw1149) [inline dnstable_convert.c:105]: I think you need to use special functions to access this variable.
From nmsg library, we use:
./nmsg/io.c:	*sum_out = atomic_load_explicit(&io->io_count_nmsg_payload_out, memory_order_relaxed);
./nmsg/io.c:		uint64_t pl_out = atomic_load_explicit(&(*io)->io_count_nmsg_payload_out, memory_order_relaxed);
./nmsg/io.c:	atomic_fetch_add_explicit(&io->io_count_nmsg_payload_out, 1, memory_order_relaxed);

**Review by djw1149 — CHANGES_REQUESTED** [review]

File: dnstable_convert.c [hunk 4]
```diff
@@ -1012,12 +1022,14 @@ init_mtbl(mtbl_compression_type compression, int level, size_t mem_mb)
 static void
 usage(const char *name)
 {
-	fprintf(stderr, "Usage: %s [-D] [-p] [-r] [-S] [-c compression] [-l level] [-m megabytes] [-s NAME] <NMSG FILE> <DB FILE> <DB DNSSEC FILE>\n", name);
+	fprintf(stderr, "Usage: %s [-D] [-p] [-r] [-S] [-c compression] [-l level] [-b size] [-t threads] [-m megabytes] [-s NAME] <NMSG FILE> <DB FILE> <DB DNSSEC FILE>\n", name);
```
Comment (shw700) [inline dnstable_convert.c]: This requires a corresponding man page entry.

Another nitpick: note the almost perfect pre-existing alphabetical ordering convention in the getopt() string itself, the subsequent option switch(), and the usage output.

File: dnstable_convert.c [hunk 5]
```diff
@@ -879,18 +988,24 @@ init_mtbl(mtbl_compression_type compression, int level, size_t mem_mb)
...
+	mtbl_writer_options_set_threadpool(wopt, pool);
+
 	mtbl_writer_options_set_compression(wopt, compression);
 	if (level != DEFAULT_COMPRESSION_LEVEL)
 		mtbl_writer_options_set_compression_level(wopt, level);
 
-	mtbl_writer_options_set_block_size(wopt, DNS_MTBL_BLOCK_SIZE);
+	if (block_size == 0)
+		mtbl_writer_options_set_block_size(wopt, DNS_MTBL_BLOCK_SIZE);
+	else mtbl_writer_options_set_block_size(wopt, block_size);
```
Comment (djw1149) [inline dnstable_convert.c]: have "else" on a line by itself

**Review by djw1149 — APPROVED** [review]

**Comment by djw1149** [comment]
Why do we only allow setting a non-default DNS data block size and not also for DNSSEC?
---
separate comment:
Maybe reword the man page text to "Use an uncompressed DNS (i.e. not DNSSEC) data block size of SIZE bytes."
because "DNS" is ambiguous here?

**Comment by kadealicious** [comment]
> ## Why do we only allow setting a non-default DNS data block size and not also for DNSSEC?
> separate comment: Maybe reword the man page text to "Use an uncompressed DNS (i.e. not DNSSEC) data block size of SIZE bytes." because "DNS" is ambiguous here?

We only allow setting a non-default DNS data block size hint because DNSSEC data blocks are typically quite bloated, so the biggest block size almost always ends up being the best option space-wise.

**Review by shw700 — APPROVED** [review]

File: dnstable_convert.c [hunk 6]
```diff
@@ -1027,17 +1024,18 @@ usage(const char *name)
...
-	" -D:       Put CDS, CDNSKEY, and TA RRSets in both outputs\n"
-	" -l LEVEL: Use numeric LEVEL of compression.\n"
-	"           Default varies based on TYPE.\n"
-	" -m MMB:   Specify maximum amount of memory to use for in-memory sorting, in megabytes.\n"
-	" -p:       Preserve empty DNS/DNSSEC files.\n"
-	" -r:       Emit RDATA and RDATA_NAME_REV dnstable entries for SOA rname field.\n"
-	" -s NAME:  NMSG source information to include in output if input is stdin.\n"
-	" -S:       Include nmsg source information in output.\n"
-	" -t COUNT: Use a maximum of COUNT threads during sorting and writing.\n");
+		" -b size:  set mtbl block size to use for DB FILE (default: %u).\n"
```
Comment (cmikk) [inline dnstable_convert.c]: The lower-casing here is incorrect, and hidden partly by the re-indent.

The metavar capitalization (e.g., `-b SIZE`) is a long-standing convention, and we should follow. Similarly, DNS rrtype mnemonics (`CDS, CDNSKEY, and TA`) are conventionally capitalized. Also, the `-D` option is upper case, and appears lowercase in the help text.

Comment (shw700) [reply dnstable_convert.c]: This is my fault. I am not sure what happened here... but it seems obvious that I accidentally lower-cased everything in vim when making the indent, and somehow missed those changes.

Comment (kadealicious) [reply dnstable_convert.c]: Pushed

File: dnstable_convert.c [hunk 7]
```diff
@@ -1058,12 +1060,19 @@ main(int argc, char **argv)
 
 		switch(c) {
 		case 'b':
-			dns_block_size = strtol(optarg, &end, 10);
-			if (*end != '\0' || dns_block_size < 1) {
-				fprintf(stderr, "Invalid DNS block size '%s'\n", optarg);
+			sscanf(optarg, "%d,%d", &dns_block_size, &dnssec_block_size);
```
Comment (cmikk) [inline dnstable_convert.c]: Using `sscanf` here is nice and succinct, but the `scanf` family of functions have a few complexities to be aware of (I tend towards "wary" of, in fact).

First, the return value of `sscanf` should be checked -- if you pass this a non-numeric string, `sscanf` will return 0, which should be treated as an error. 

Second, even checking the value of `sscanf` will miss `-b 65536,blah` -- `sscanf` will collect the 65536, return 1, and ignore the `,blah`.  You could either:
1. separately check the string for a `,` and require `sscanf` to return 2 if a `,` is present, or
2. do a second `strtol` call if the first ends with `*end == ','`. This also simplifies error reporting of invalid dns vs dnssec block size

Finally, the dnssec block size is an unsigned `size_t` (and the variables should be, too) -- if you stick with `sscanf`, use `%zu` for the format; if using `strtoXX` functions, use `strtoul`.

File: dnstable_convert.c [hunk 8]
```diff
@@ -1031,16 +1045,40 @@ main(int argc, char **argv)
...
 
-	while ((c = getopt(argc, argv, "Dc:l:m:prSs:")) != -1) {
+	while ((c = getopt(argc, argv, "b:Dc:l:m:t:prSs:")) != -1) {
 		mtbl_res res;
 		char *end;
 
 		switch(c) {
+		case 'b':
+			if (sscanf(optarg, "%d,%d", &dns_block_size, &dnssec_block_size) != 2 &&
+			    strchr(optarg, ',') != NULL) {
```
Comment (cmikk) [inline dnstable_convert.c]: This will still silently ignore an invalid number without a ','.  The full `sscanf` treatmet would be:
```
if (strchr(optarg, ',') != NULL) {
     if (sscanf(optarg, "%d,%d", ...) != 2) {
           /* error */
     }
} else if (sscanf(optarg, "%d", ...) != 1) {
     /* error */
}
```

I would suggest considering `strtol` again, like
```
dns_block_size = strtol(optarg, &endp, 10);
switch(*endp) {
default: 
     /* malformed block size */
     return (EXIT_FAILURE);
case '\0': break;
case ',': 
      optarg = endp;
      dnssec_block_size = strtol(optarg, &endp, 10);
      if (endp != '\0') {
           /*  malformed dnssec block size */
           return (EXIT_FAILURE);
     }
}
``` 

(Note that the `strtol` approach allows us to easily support `N,M`, `N:M`, `N/M` with case fallthrough.)

Comment (kadealicious) [reply dnstable_convert.c]: I am confused here.  When I run this with an invalid block size (e.g. -1, 0, -22) with or without a comma, the function does indeed return `EXIT_FAILURE`.  An invalid number for either block size is anything less than 1, which are the if statements that is immediately follow the initial `sscanf()`.

Comment (cmikk) [reply dnstable_convert.c]: You are correct when the input parses as a number, but is out of range (i.e., zero or negative).  If the `-b` argument doesn't parse as a number, `sscanf` returns 0, and the current logic will only reject the argument if it contains a comma.

Comment (kadealicious) [reply dnstable_convert.c]: Revisited the `strtoul()` approach, works great now!

**Review by cmikk — APPROVED** [review]
